crates/noirc_frontend/src/hir/type_check/expr.rs

```diff
@@ -229,113 +211,198 @@ pub(crate) fn type_check_expression(
-                    FieldElementType::Private,
-                    ArraySize::Variable,
-                    Box::new(last_type),
-                ),
-            );
+            Type::Array(
+                // The type is assumed to be private unless the user specifies
+                // that they want to make it public on the LHS with type annotations
+                FieldElementType::Private,
+                ArraySize::Variable,
+                Box::new(last_type),
+            )
         }
         HirExpression::Block(block_expr) => {
             for stmt in block_expr.statements() {
                 super::stmt::type_check(interner, stmt)?
             }
-            let last_stmt_type = match block_expr.statements().last() {
+
+            match block_expr.statements().last() {
                 None => Type::Unit,
                 Some(stmt) => extract_ret_type(interner, stmt),
-            };
-
-            interner.push_expr_type(expr_id, last_stmt_type)
+            }
         }
         HirExpression::Prefix(_) => {
             // type_of(prefix_expr) == type_of(rhs_expression)
             todo!("prefix expressions have not been implemented yet")
         }
-        HirExpression::Predicate(_) => {
-            todo!("predicate statements have not been implemented yet")
+        HirExpression::If(if_expr) => {
+            let cond_type = type_check_expression(interner, &if_expr.condition)?;
+            let then_type = type_check_expression(interner, &if_expr.consequence)?;
+
+            if cond_type != Type::Bool {
+                return Err(TypeCheckError::TypeMismatch {
+                    expected_typ: Type::Bool.to_string(),
+                    expr_typ: cond_type.to_string(),
+                    expr_span: interner.expr_span(&if_expr.condition),
+                });
+            }
+
+            match if_expr.alternative {
+                None => Type::Unit,
+                Some(alternative) => {
+                    let else_type = type_check_expression(interner, &alternative)?;
+
+                    if then_type != else_type {
+                        let mut err = TypeCheckError::TypeMismatch {
+                            expected_typ: then_type.to_string(),
+                            expr_typ: else_type.to_string(),
+                            expr_span: interner.expr_span(&alternative),
+                        };
+
+                        if then_type == Type::Unit {
+                            err = err
+                                .add_context(
+                                    "Are you missing a semicolon at the end of your 'else' branch?",
+                                )
+                                .unwrap();
+                        } else if else_type == Type::Unit {
+                            err = err
+                                .add_context(
+                                    "Are you missing a semicolon at the end of your 'then' branch?",
```

Comment:
What does the 'then' branch refer to?

Reply:
It refers to the expression executed if the condition is true which in noir terms is the consequence. I had never heard of consequence/alternative used for these branches before so I used then/else here for user facing errors since I thought they'd be more common.

I just did a quick survey of what wording other languages give for this error.
Rust: `error[E0308]: `if` and `else` have incompatible types`
GHC, OCaml, and Purescript give generic type errors.
Elm gives `The 2nd branch of this `if` does not match all the previous branches: ...`

Maybe we should change it to be more generic?: "Are you missing a semicolon at the end of this branch?" while pointing to the 'then' branch.

Reply:
Yep that would be amazing, but not blocking for this PR. For now , can we change 'then' to 'else' because the syntax itself does not have 'then' IIRC